chain/client/src/stateless_validation/chunk_endorsement/tracker_v2.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::HashMap;
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Mutex};
+
+use lru::LruCache;
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::stateless_validation::chunk_endorsement::ChunkEndorsementV2;
+use near_primitives::stateless_validation::ChunkProductionKey;
+use near_primitives::types::AccountId;
+use near_primitives::version::ProtocolFeature;
+use near_store::Store;
+
+use crate::stateless_validation::validate::validate_chunk_endorsement;
+
+use super::ChunkEndorsementsState;
+
+// This is the number of unique chunks for which we would track the chunk endorsements.
+// Ideally, we should not be processing more than num_shards chunks at a time.
+const NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE: usize = 100;
+
+/// Module to track chunk endorsements received from chunk validators.
+pub struct ChunkEndorsementTracker {
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// Used to find the chain HEAD when validating partial witnesses.
+    store: Store,
+    /// We store the validated chunk endorsements received from chunk validators.
+    /// Interior mutability is required to update the cache in process_chunk_endorsement & compute_chunk_endorsements.
+    chunk_endorsements: Mutex<LruCache<ChunkProductionKey, HashMap<AccountId, ChunkEndorsementV2>>>,
+}
+
+impl ChunkEndorsementTracker {
+    pub(crate) fn new(epoch_manager: Arc<dyn EpochManagerAdapter>, store: Store) -> Self {
+        Self {
+            epoch_manager,
+            store,
+            chunk_endorsements: Mutex::new(LruCache::new(
+                NonZeroUsize::new(NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE).unwrap(),
+            )),
+        }
+    }
+
+    // Validate the chunk endorsement and store it in the cache.
+    pub(crate) fn process_chunk_endorsement(
+        &self,
+        endorsement: ChunkEndorsementV2,
+    ) -> Result<(), Error> {
+        // Check if we have already received chunk endorsement from this validator.
+        let key = endorsement.chunk_production_key();
+        let account_id = endorsement.account_id();
+        if self
+            .chunk_endorsements
+            .lock()
+            .unwrap()
+            .peek(&key)
+            .is_some_and(|entry| entry.contains_key(account_id))
+        {
+            tracing::debug!(target: "client", ?endorsement, "Already received chunk endorsement.");
+            return Ok(());
+        }
+
+        // Validate the chunk endorsement and store it in the cache.
+        if validate_chunk_endorsement(self.epoch_manager.as_ref(), &endorsement, &self.store)? {
+            let mut chunk_endorsements = self.chunk_endorsements.lock().unwrap();
```

Comment:
nit: having temporary variables seems redundant here and also you don't need to clone `account_id`:
```
self.chunk_endorsements.lock().unwrap()
    .get_or_insert_mut(key, || HashMap::new())
    .insert(account_id, endorsement);
```

Reply:
I unfortunately can't do this in the second location as rust complains `temporary value dropped while borrowed`